purenix/src/Nix/Convert.hs

```diff
@@ -0,0 +1,174 @@
+            mempty
+        )
+  where
+    inheritFrom :: P.ModuleName -> [Ident] -> Convert (N.Expr, [N.Ident])
+    inheritFrom (P.ModuleName m) exps = (N.sel (N.var "modules") m,) <$> traverse ident exps
+
+    foreignBinding :: Ident -> Convert (N.Ident, N.Expr)
+    foreignBinding ffiIdent = do
+      i <- ident ffiIdent
+      pure (i, N.sel (N.var "__ffi") i)
+
+bindings :: [Bind Ann] -> Convert [(N.Ident, N.Expr)]
+bindings = traverse binding . (>>= flatten)
+  where
+    binding :: (Ann, Ident, Expr Ann) -> Convert (N.Ident, N.Expr)
+    binding (ann, i, e) = localAnn ann $ liftA2 (,) (ident i >>= checkKeyword) (expr e)
+    flatten :: Bind a -> [(a, Ident, Expr a)]
+    flatten (NonRec a i e) = [(a, i, e)]
+    flatten (Rec bs) = (\((a, i), e) -> (a, i, e)) <$> bs
+
+expr :: Expr Ann -> Convert N.Expr
+expr (Abs ann arg body) = localAnn ann $ liftA2 N.abs (ident arg >>= checkKeyword) (expr body)
+expr (Literal ann lit) = localAnn ann $ literal lit
+expr (App ann f x) = localAnn ann $ liftA2 N.app (expr f) (expr x)
+expr (Var ann (P.Qualified Nothing i)) = localAnn ann $ N.var <$> ident i
+expr (Var ann (P.Qualified (Just (P.ModuleName m)) i)) = localAnn ann $ N.sel (N.sel (N.var "modules") m) <$> ident i
+expr (Accessor ann sel body) = localAnn ann $ flip N.sel (removeQuotes $ P.prettyPrintObjectKey sel) <$> expr body
+expr (Let ann binds body) = localAnn ann $ liftA2 N.let' (bindings binds) (expr body)
+expr (ObjectUpdate ann a b) = localAnn ann $ liftA2 (N.bin N.Update) (expr a) (attrs b)
+expr Case {} = throw "Cannot yet convert case expression"
+expr Constructor {} = throw "Cannot yet convert constructors"
+
+ident :: Ident -> Convert N.Ident
+ident (Ident i) = pure i
+-- GenIdent is only used in PureScript for "unnamed" instances.
+-- Originally, in PureScript, all instances needed to be named:
+-- https://github.com/purescript/documentation/blob/master/language/Differences-from-Haskell.md#named-instances
+-- This was relaxed in 0.14.2:
+-- https://github.com/purescript/purescript/pull/4096
+-- TODO: We'll have to make sure that no identifier are created that are _only_
+-- an integer (when mname is Nothing), since they can't be used in Nix.
+ident (GenIdent mname n) = pure $ maybe id mappend mname (T.pack $ show n)
+ident UnusedIdent = throw "Impossible: Encountered typechecking-only identifier"
+
+checkKeyword :: N.Ident -> Convert N.Ident
+checkKeyword w
+  | w `elem` purenixIdents = throw $ "binder " <> w <> " is a special identifier in purenix"
+  | w `elem` nixKeywords = throw $ "binder " <> w <> " is a nix keyword"
+  | w `elem` nixPrimops = throw $ "binder " <> w <> " is a nix primop.  You probably don't want to shadow this."
+  | otherwise = pure w
+  where
+    -- These idents have a special meaning in purenix.
+    purenixIdents = ["modules", "__ffi"]
+    -- primops (builtins) in Nix that can be accessed without importing anything.
+    -- These were discovered by running `nix repl` and hitting TAB.
+    nixPrimops =
```

Comment:
At this point it probably makes sense to turn this into a `Set` for cheap lookups

Reply:
Yeah, that completely makes sense.